apiserver/facades/controller/charmrevisionupdater/state_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmrevisionupdater_test
+
+import (
+	"time"
+
+	"github.com/juju/charm/v8"
+	"github.com/juju/charm/v8/resource"
+	csparams "github.com/juju/charmrepo/v6/csclient/params"
+	"github.com/juju/collections/set"
+	"github.com/juju/names/v4"
+
+	"github.com/juju/juju/apiserver/facades/controller/charmrevisionupdater"
+	"github.com/juju/juju/cloud"
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/state"
+)
+
+type mockState struct {
```

Comment:
Please use go mock for these, it is the preferred juju unit test pattern.  Overtime time this style of mocking becomes limited in what it can do.  We'd only use this style for something like a logger.  As well, we cannot verify inputs easily with it, nor have tests with different results for easy error scenarios.